Added filesize options, clustering check, change mode and timestamp use-case, keep newest attributes.

wolfospealain opened the pull request.

wolfospealain committed:
Correct print statement bug.

wolfospealain committed:
Add file size options.

Memory can be a problem over a large number of files. Added file-size and max-file-size to check only files of a certain size min, max or range. It is ultimately more important to check all identical files in a directory than check all files and run short on memory. It also provides a quick way to save space by hardlinking only the larger files.

wolfospealain committed:
Fix clustering problem.

There is a problem when we meet two files that already have hardlinks each. It is possible for hardlink.py to idly move a file from one hardlink to another without ever ultimately saving space. This can happen on large disks with backup files, stored and occasionally hardlinked over years. The solution is to always use the file with the most hardlinks as the source if (stat_info.st_nlink > temp_stat_info.st_nlink).

wolfospealain committed:
Replace --content-only and --notimestamp.

Replace --content-only with --properties and replace --notimestamp with --timestamp which default to false - more consistent with the hardlink.py space saving use-case.

wolfospealain committed:
Verbosity changes.

wolfospealain committed:
Minor format edits.

wolfospealain committed:
Keep newest attributes.

When creating hardlinks, keep the attributes of the newest file.

jamescassell commented:
Thanks for the much-improved pull request!  I'm not sure if @akaihola has had a chance to look at it yet.  As for me, I appreciate the detailed explanations of what is changed in each commit.  I'm happy with all the changes, except that I'll defer to @akaihola on the defaults changes for "more consistent with the hardlink.py space saving use-case"

Thanks for the well-formed PR!

akaihola commented on hardlink.py:
```diff
@@ -66,8 +66,8 @@ def hash_size(size):
-def hash_value(size, time, notimestamp):
-    if notimestamp:
+def hash_value(size, time, timestamp):
+    if not(timestamp):
```
This doesn't look like Python. The proper way to write a negative boolean test is:

    if not timestamp:

The same applies to lines 101, 104, 107, 110 and 111 below.

akaihola reviewed (changes requested):
I like the changes, and I think they do improve usability.

Since there are multiple backwards incompatible changes in command line options, I would bump up the version number. Since we're not yet at 1.0, it would probably be appropriate to go to 0.6 now.

akaihola commented on hardlink.py:
```diff
@@ -382,17 +391,21 @@ def parse_command_line():
     parser.add_option("-q", "--no-stats", help="Do not print the statistics",
                       action="store_false", dest="printstats", default=True,)
 
-    parser.add_option("-t", "--timestamp-ignore",
```
Removal of this command line option breaks the `test_hardlink_tree_timestamp_ignore` test in `tests.py`:
```python
test_hardlink_tree_timestamp_ignore (tests.TestHappy) ...
Usage: hardlink.py [options] directory [ directory ... ]

hardlink.py: error: no such option: --timestamp-ignore
ERROR

======================================================================
ERROR: test_hardlink_tree_timestamp_ignore (tests.TestHappy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "hardlinkpy/tests.py", line 124, in test_hardlink_tree_timestamp_ignore
    hardlink.main()
  File "hardlinkpy/hardlink.py", line 457, in main
    options, directories = parse_command_line()
  File "hardlinkpy/hardlink.py", line 422, in parse_command_line
    (options, args) = parser.parse_args()
  File "/usr/lib64/python2.7/optparse.py", line 1402, in parse_args
    self.error(str(err))
  File "/usr/lib64/python2.7/optparse.py", line 1584, in error
    self.exit(2, "%s: error: %s\n" % (self.get_prog_name(), msg))
  File "/usr/lib64/python2.7/optparse.py", line 1574, in exit
    sys.exit(status)
SystemExit: 2
```

wolfospealain replied:
Updated test at https://github.com/akaihola/hardlinkpy/pull/3/commits/0abbf373bc02bcd26db7d25ea067e2ae581d72cc to test for the alternative.

akaihola commented:
Commit 042b0cb introduces the following test failures:
```python
======================================================================
FAIL: test_hardlink_tree (tests.TestHappy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "hardlinkpy/tests.py", line 92, in test_hardlink_tree
    self.assertNotEqual(get_inode("dir1/name1.ext"), get_inode("dir4/name1.ext"))
AssertionError: 777571 == 777571

======================================================================
FAIL: test_hardlink_tree_exclude (tests.TestHappy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "hardlinkpy/tests.py", line 120, in test_hardlink_tree_exclude
    self.assertNotEqual(get_inode("dir1/name1.ext"), get_inode("dir4/name1.ext"))
AssertionError: 777597 == 777597

======================================================================
FAIL: test_hardlink_tree_filenames_equal (tests.TestHappy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "hardlinkpy/tests.py", line 101, in test_hardlink_tree_filenames_equal
    self.assertEqual(get_inode("dir1/name1.ext"), get_inode("dir2/name1.ext"))
AssertionError: 777610 != 777611

======================================================================
FAIL: test_hardlink_tree_match (tests.TestHappy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "hardlinkpy/tests.py", line 148, in test_hardlink_tree_match
    self.assertNotEqual(get_inode("dir1/name1.ext"), get_inode("dir4/name1.ext"))
AssertionError: 777623 == 777623
```

wolfospealain commented:
No timestamp checking is now the default, hence the failure of the tests which assumed timestamp checking. Adding the option "-T" to the test options causes the original behaviour and OK tests.

wolfospealain committed:
Update tests for new default behaviour.

wolfospealain committed:
Correction to Python style.

wolfospealain commented:
New defaults, solution to the clustering problem, and the retention of the lastest attributes are working well with Back In Time backups - reclaiming more space and ensuring future backups are minimal (rsync will now compare candidate  files with the latest file attributes in the backup).

wolfospealain committed:
Maximum Hardlinks

Check if a file has already reached maximum hard links before attempting to hardlink.

wolfospealain committed:
Humanise output.

Standardise B, KiB, MiB, GiB.

akaihola commented on hardlink.py:
```diff
@@ -225,9 +225,11 @@ def hardlink_identical_files(directories, filename, options):
-                (stat_info[stat.ST_SIZE] <= options.max_file_size) or (options.max_file_size == 0)):
+    # Is it a regular file, within size limits, and with less than the maximum hardlinks already?
+    elif stat.S_ISREG(stat_info[stat.ST_MODE]) \
+            and (stat_info[stat.ST_SIZE] >= options.file_size) \
```
This and the line two lines below have redundant parentheses.

You could just write:
```python
elif stat.S_ISREG(stat_info[stat.ST_MODE]) \
    and stat_info[stat.ST_SIZE] >= options.file_size \
    and (stat_info[stat.ST_SIZE] <= options.max_file_size
         or options.max_file_size == 0) \
    and stat_info.st_nlink < os.pathconf(filename, "PC_LINK_MAX"):
```

Many Pythonistas also like to use parentheses instead of backslash line continuations:

```python
elif (stat.S_ISREG(stat_info[stat.ST_MODE])
      and stat_info[stat.ST_SIZE] >= options.file_size
      and (stat_info[stat.ST_SIZE] <= options.max_file_size
           or options.max_file_size == 0)
      and stat_info.st_nlink < os.pathconf(filename, "PC_LINK_MAX")):
```

chadnetzer commented on hardlink.py:
```diff
@@ -225,9 +225,11 @@ def hardlink_identical_files(directories, filename, options):
+    elif stat.S_ISREG(stat_info[stat.ST_MODE]) \
+            and (stat_info[stat.ST_SIZE] >= options.file_size) \
+            and ((stat_info[stat.ST_SIZE] <= options.max_file_size) or (options.max_file_size == 0)) \
+            and (stat_info.st_nlink < os.pathconf(filename, "PC_LINK_MAX")):
```
Calling os.pathconf() per-filename means (potentially) multiple os syscalls (stat/access/readlink); this appears to be true for Linux ext4 and glibc, for example.  This could have a non-negligible performance effect.

Since the condition of reaching max nlinks is likely rare, it might be easier and quicker just to avoid the pre-check.

akaihola replied:
I like the idea of avoiding the pre-check and catching the error. I wonder if the exception raised is consistent across operating systems and filesystems, though. Any idea?

chadnetzer replied:
Current code catches any exception during os.link() and treats as an error, so likely no changes need to be made.  The os.rename() exception handling is slightly inconsistent, but acceptable (imo).

On Python2, it appears os.link() can return a WindowsError on Windows, whereas on Posix it'll return an OSError for operating system errors.  In Python3, WindowsError has been merged into OSError, so it should be easier to deal with.  There can also be TypeErrors or NotImplementedErrors for bogus arguments to os.link() and other functions, such as None.

chadnetzer reviewed (commented):
This change is desirable, but unfortunately breaks the tests (on some systems) due to issues similar to #5.  The current algorithm doesn't ensure all identical files are hardlinked, and the tests are written assuming specific files will end up hardlinked.  This is really a problem with the algorithm first and the tests second, but this patch exacerbates the problem.

akaihola commented:
So @chadnetzer do I understand correctly that the modified (or also original?) algorithm isn't perfectly consistent across different operating systems?

chadnetzer commented:
@akaihola - Yes, or at least it's dependent on the order that os.listdir() returns directory items, which can vary by OS, filesystem, etc.  Basically by hardlinking to a file which already has more than 1 hardlink, the old destination hardlink is replaced by a new different hardlink.  Since we are literally attempting to hardlink all the identical files, this property of linking works against us, and identical files can still end up unlinked because of it; also it makes the tests, which are checking for results based on a specific order, unreliable.

I think the best solution is to work on an updated algorithm as suggested in the TODO section at the top of hardlink.py, and gather up all the information on all the files *before* any hardlinking happens.  We basically want to apriori discover which sets of files should be linked together, and then perform the linking in the correct order so as to ensure they all remain linked together.  Furthermore, we may also want to retain information on which files are already linked together, to avoid breaking those existing links (such as when we are in the "filenames-equal" mode).

FWIW, I'm starting to work on the TODO version of the algorithm, which I expect will ensure that *all* identical files will end up hardlinked together, which should also fix the issue with the tests.

akaihola commented:
@chadnetzer, I'm struggling to understand the problem of the algorithm fully. Am I grasping it correctly so far:

- `os.listdir()` returns entries in a non-deterministic order
- given no existing hard links, the end result is deterministic (barring unexpected termination)
- even with existing hard links, the end result is deterministic (barring unexpected termination)
- after making the first new hard link, and before the last new hard link is made, the set of hard links at each step of the algorithm is dependent on `os.listdir()` behavior
- given no pre-existing hard links, the algorithm won't break any hard links it has made along the way
- the algorithm may break some pre-existing hard links by re-linking to a different identical file (but the end result will still be the same)?

So for example, if we have three identical files A, B, and C, with B and C already hardlinked, the algorithm would first register A:

    A BC
    ^

the algorithm would then register B, noticing it's identical to A. It would hard link B to A, breaking the link between B and C:

    AB C
     ^

Finally, it would register C and link it together with A:

    ABC
      ^

Am I correct that the problem you describe only has two ill effects:
- sub-optimal performance due to "going back and forth" re-linking existing links
- non-deterministic order of processing, resulting in
  - complications in the test suite
  - unpredictable end state when execution terminates unexpectedly?

akaihola commented:
Also @chadnetzer, is the memory footprint going to stay the same even if all directories are scanned upfront before resolving the new hard links to be made?

On I quick glance I would expect so, because even in the current "incremental" mode, information about every encountered file is still eventually held in memory simultaneously.

chadnetzer commented:
@akaihola I'm responding now even though I still have some investigating to do, because we have a holiday coming up and I didn't want to keep you waiting.

Thanks for the well described algorithm breakdown with examples.  I think you are mostly correct.  The current algorithm as it is (with an exception I'll note below), works consistently in part because it keeps track of files it has seen, and always uses a previously seen file as the hardlink source.  Which means that as it goes along, even though it may "break" links for newly discovered files, it (I think) always ends up links of the files it has already seen.  So that when the algorithm is finished, it has linked all the identical files to a common inode.  BTW, I think this is why the "clustering fix" commit from @wolfospealain changes the results (and causes more test failures), as it can reverse the link ordering, and thus break links between files that are already seen (ie. in the file_hashes dict), in favor of not breaking a link on a newly seen file.  While this could lead (overall) to fewer relinks, it also (apparently) means there can be a different end result of which files are linked.  This may or may not be easily changed behavior, I don't know yet.

The big caveat to your analysis is that, with the "--filenames-equal" option selected, the algorithm doesn't work as you described.  In that case, a newly discovered file may have an identical hash but different basename, and *not* end up linked to an already seen file.  If it is also linked to other files, which do have identical filenames, those links end up getting broken.  Ie. in that case, when the algorithm terminates, files that were previously linked can definitely end up unlinked, because they have different filenames.

This may be desired behavior, in fact.  It's sort of a policy question about whether the "--filenames-equal" option implies that existing hardlinks between differently named files are meant to be preserved after the program completes.  It seems somewhat non-trivial to ensure that behavior (and in any case, nothing can be done about preserving hardlinks to files that are in other directories on the filesystem which are not supplied as arguments to the program; ie. the program can't ensure preservation of hardlinks in all cases).  So, I think it's perfectly reasonable to allow the "--filenames-equal" option to break already existing links to unequally named files.  Note that there *is* a bug even with this option, that was causing tests to fail, because the is_already_hardlinked() test doesn't consider filenames (but filenames matter when the option is selected).  So, that behavior is addressed in my PR #11.

So I think generally what you've stated is correct (with that caveat about the equal filenames option), and that the current algorithm does a good job on the whole.  As you indicate, changing the algorithm to try and minimize link breaking may require using more memory, for perhaps little gain.

It's really the current testing methodology that is sensitive to changes in things like os.listdir() order, or changes to linking order, etc.  I may spend some time on thinking about ways to improve the testing; ideally it seems we may want not to test for specific links between files, but overall results (ie. did no unequal files get linked, did we actually save space, etc.).

chadnetzer reviewed (commented):
This "Keep newest attributes" commit doesn't achieve what is intended, afaict.  The update_stat_info is passed into hardlink_files() which only uses the stat_info object for printing output.  Once two files are linked, they have the same inode, and thus the same stat_info anyway.

I assume the intent was to choose to retain the inode with the newest st_mtime?  If so, it is possible to set the mtime of the link inode to the newest value after the link is performed (like "touch" does).

wolfospealain commented:
The cluster fix keeps the inode with the most links - reducing the possibility of separating links, and as an incremental procedure across many reuses of the program it eliminates future unnecessary breaking of links. Only the attributes are changed to match the new st_mtime - which keeps the process compatible with archiving tools such as backintime and rysnc generally.

chadnetzer commented:
@wolfospealain Unfortunately, I think the clustering fix as written is buggy.  Thanks belong to @akaihola for helping me understand this better.

In the original algorithm, the hardlink source file is in the file_hashes dict, where the values are (pathname, stat_info) pairs.  Because it is the hardlink source, it keeps the same inode after the hardlink.  However, by reversing the hardlink order, it is now the destination, which means it gets a new inode number, but the stat_info in file_hashes isn't updated and has the old st_ino and st_nlinks data.  This will affect subsequent is_already_hardlinked() calls, as there will now potentially be false positives and some later files could get skipped.

I'll use @akaihola notation to explain what I think is happening.  Let's say we have files A and E which are linked together, and files B, C, and D which are linked together (all identical and thus hardlinkable).  Let's assume the tree walk discovers them in alphabetical order (ie. A, B, C, D, E).  In the current algorithm, the grouping/linking will progress as:

    AE   BCD  # A is seen, A added to file_hashes
    ABE   CD  # B is seen, B gets linked to A
    ABCE   D  # C is seen, C gets linked to A
    ABCDE     # D is seen, D gets linked to A
    ABCDE     # E is seen, already linked, skipped


With the clustering fix, as it currently works, the linking will occur like this:

    AE   BCD  # A is seen, A added to file_hashes
    ABCD   E  # B is seen, A linked to BCD cluster.  A file_hashes data still has E's inode number
    ABCD   E  # C is seen, A gets re-linked to C (file_hashes nlink data is stale)
    ABCD   E  # D is seen, A get re-re-linked to D
    ABCD   E  # E is seen, file_hashes says it is already linked A, skipped


So in this scenario (if I'm right), a linkable file has been skipped, and some redundant re-links have occurred.  For this clustering fix to work, it'll have to properly update the state_info data in the file_hashes cache (st_ino and st_nlinks, in particular).  There may be other caveats, I'll have to think about it a bit more.  It's possible I haven't analyzed it correctly, but what we do know is that with the clustering fix, previously passing tests are failing.

Regarding the mtime, you say the attributes are changed to match the new st_mtime, but I don't see that actually occurring. The stat_info with the newer mtime is passed to hardlink_files, which doesn't do anything with it further.  If the intent was for the linked inode to be updated to the more recent mtime, that isn't being done afaict.

wolfospealain committed:
Missing code to update attributes to newest.

wolfospealain committed:
Update file hashes following hardlink

wolfospealain commented:
Well spotted. File hashes needws to be updated following any hardlink - to ensure the correct inode and the updated links count and attributes.

chadnetzer commented on hardlink.py:
```diff
@@ -162,6 +162,9 @@ def hardlink_files(sourcefile, destfile, stat_info, options):
             if not options.dryrun:
                 os.link(sourcefile, destfile)
+                # Update destination to latest attributes
+                os.chown(destfile, stat_info.st_uid, stat_info.st_gid)
```
It's possible for the link() to succeed, and the chown() to fail (such as when sourcefile owner doesn't match the user). This will trigger the link() unwind behavior, which will itself fail, because the link() actually succeeded.

So the chown() and utime() need to be in their own try/except, probably somewhere in the "link succeeded" section. Presumably, if the chown() or utime() fail, the link should still remain. Also, the chown() and utime() aren't needed if the stat_info already matches the sourcefile's, and it'd be beneficial IMO not to unconditionally trigger two additional inode updates per file if they aren't always necessary.

Finally, although it may be tricky to develop, it'd be nice to have a test case that would've demonstrate this change failing (and passing when it's fixed).

wolfospealain committed:
Update file hashes (correction for failed link).

wolfospealain committed:
Only update file attributes if necessary.

chadnetzer commented on hardlink.py:
```diff
@@ -175,6 +172,10 @@ def hardlink_files(sourcefile, destfile, stat_info, options):
             # hard link succeeded
+            # Update destination to latest attributes
+            if update:
+                os.chown(destfile, stat_info.st_uid, stat_info.st_gid)
```
Again, this could throw an uncaught exception when the link() was created from a file not owned by the user, which isn't an unexpected condition in cases where (say) multiple users on a system may be consolidating downloaded package repositories, etc.

wolfospealain replied:
Should I have permission to create links from a file where I don't have write permissions anyway? It is a much larger question as to whether creating hardlinks intended for routine use by different users is ever appropriate unless running as root to consolidate archives and assume ownership. A hardlink can only have one owner and one set of attributes - either way someone is going to lose ownership. The -P option is there to decide on the particular use case intended.

wolfospealain replied:
Declustering will still not be achieved in a single pass - depending on whether files from smaller clusters are encountered before encountering larger clusters of identical hardlinkable files. Inevitable when analysing and changing files on the fly. However redundant re-linking is now avoided. Clusters will consolidate on repeated passes - saving additional space.

chadnetzer replied:
Should I have permission to create links from a file where I don't have write permissions anyway?

Sure.  Certainly it shouldn't crash the program.

wolfospealain replied:
True. But if I correctly created a link what would stop the editing of ownership and time?

chadnetzer replied:
The fact that you don't own the file.

You control the directory entry, so you could delete the link.  But that's about it.  Of course, you always have the option to un-consolidate the file by copying it.

wolfospealain committed:
Catch chown/utime exceptions.

wolfospealain committed:
Maximum Links

Check for maximum links, once, on first directory. Hardlinks are not made between different filesystems anyway. Check maximum links before including  a new and while testing is two files are hardlinkable. It is a high number (Linux 65,000) but it can easily be met while hardlinking archives of files.

chadnetzer commented on hardlink.py:
```diff
@@ -236,7 +238,7 @@ def hardlink_identical_files(directories, filename, options):
             and (stat_info[stat.ST_SIZE] >= options.file_size) \
             and ((stat_info[stat.ST_SIZE] <= options.max_file_size) or (options.max_file_size == 0)) \
-            and (stat_info.st_nlink < os.pathconf(filename, "PC_LINK_MAX")):
+            and (stat_info.st_nlink < maximum_links:
```
Syntax Error

Also, it may be better not to link a pair of files if their total st_nlink adds up to more than the maximum_links.  This is because, in this situation, no space saving is possible.  Linking two existing files preserves the total nlink count, so neither of the two inodes will go to zero even if all it's filenames are found.  Skipping such pairs avoids unnecessary inode changes.

wolfospealain committed:
# bad copy & paste skills

chadnetzer commented:
@wolfospealain The "clustering fix" still has a problem; since a file that has already been processed and is in file_hashes can now be the destination link, it gets unlinked from the other files that it was linked to which came before it.  Because the algorithm is single pass this means there can be left-over identical files when processing has finished.  To work as you want, all the files other linked to the destination file that is in file_hashes would also have to be linked to the larger incoming "cluster".

I wrote a test case that demonstrates this problem on my devel branch at commit 70773c1

wolfospealain commented:
True, it doesn't complete the task in one pass. But at least as a patch it doesn't make things worse. A rewrite of the algorithm capable of backtracking through previous occurrences of an inode is required.

wolfospealain committed:
Create LICENSE

wolfospealain committed:
Delete COPYING

wolfospealain committed:
Delete setup.py

wolfospealain committed:
Delete .gitignore

wolfospealain committed:
Rewrite as Python3 OO code. Restructured database for backtracking.

Python 3.
Rewritten/refactored as OO code.
New database structure to effectively handle clusters and backtracking, with integrated statistics.
Added debug option for effective database debugging, testing, development.
Added install option.
Improved output formatting.
Switched to os.scandir.
Switched to argparse.
Removed mirror.pl and rsync exclusions (up to user to exclude files).
Eliminated some redundant eligibility checks and duplicate is_already_hardlinked.
Consolidated code functions.
Cleared legacy debug code.
Removal of unnecessary max_hashes.
Added new option for hardlinks summary output (default: false).
Removed output sorting.

wolfospealain committed:
Additional tests. Removal of temporary files.

Additional tests. Change to test data generation. Swtich to using tempfile.TemporaryDirectory().

wolfospealain committed:
Minor edit.

wolfospealain committed:
Update AUTHORS.rst

wolfospealain committed:
Python 3

wolfospealain committed:
Delete AUTHORS.rst

wolfospealain committed:
Delete README.rst

wolfospealain committed:
Create README.md

wolfospealain committed:
Comment updates.

wolfospealain committed:
Update hardlink.py

wolfospealain committed:
Update README.md

wolfospealain committed:
Create hardlink_profile_code.py

wolfospealain committed:
Update hardlink.py

wolfospealain committed:
Rename hardlink_profile_code.py to _profile.py

wolfospealain committed:
Rename tests.py to _tests.py

wolfospealain committed:
Update hardlink.py

wolfospealain committed:
Update hardlink.py

chadnetzer commented:
@wolfospealain The hardlink() method of your current master branch needs "dry-run" checking around the chown()/utime() updates, as it currently attempts to modify the filesystem even in dry-run mode.

Best and safest solution, imo, is to put dry-run check once at the beginning of the method, rather than doing the check repeatedly throughout the method.

akaihola commented on hardlink.py:
```diff
@@ -316,7 +316,7 @@ def human(number):
 def parse_command_line(version, install_path):
     description = "%(prog)s version " + version + ". " \
-                  + "Scan for and hardlink identical files."
+                  + "Scan for and hardlink identical files. https://github.com/wolfospealain/hardlink"
```
Did you intend this change only for your own fork, or to be merged into other forks as well?

Before making this change, I suggest that we discuss and reach an agreement between active `hardlinkpy` contributors about which GitHub repository should be the one we want to point users to.

Or, would it be preferable to point to a PyPI project instead? Currently [pypi.org/project/hardlinkpy](https://pypi.org/project/hardlinkpy/) is owned by @JohnVillalovos, and links to his initial export from Google Code at [github.com/JohnVillalovos/hardlinkpy](https://github.com/JohnVillalovos/hardlinkpy) as the project home page.

We could
- create an organization and project on GitHub separate from our personal accounts, or
- rename the project and upload it to PyPI, or
- ask if @JohnVillalovos is willing to step in and collaborate with us, or
- ask if @JohnVillalovos is willing to hand over his PyPI entry.

If we only pointed to a PyPI project in the code and documentation, future changes to the "official" GitHub repository wouldn't need any changes in the code base.

Ping @chadnetzer, your opinion?

akaihola commented:
Regarding splitting this up into multiple pull requests, I noticed that in his [comment](/akaihola/hardlinkpy/pull/2#pullrequestreview-101641077) to #2, @jamescassell wrote a detailed suggestion for changes up to that point. I assume #2 (now closed) was later turned into #3 (this pull request).

JohnVillalovos commented:
Hi there,

I am currently on vacation but will be back by 1-Aug-2018. I'll try to be more responsive then.

Thanks,
John

wolfospealain committed:
Fix hardlink() dry_run.

chadnetzer commented:
@akaihola @wolfospealain @jamescassell I suggest opening an issue to discuss future of hardlinkpy development, and some of the comments of @akaihola raises in his comment statement from earlier.

chadnetzer commented:
@wolfospealain I think it's worth closing this PR unmerged, since as you note, it has diverged from the upstream master significantly since it's original approval months ago.  However, if you are willing, you can comment in issue #36 if you have any thoughts you wish to add to the ongoing development of this repo.  It appears that you have decided to maintain your own repo fork independently (I'm guessing?), and I have also been working on my variant which has some similar goals (although with very different implementations), and it might be worth a general discussion about where we all want to end up, and what benefits (if any) might be had by collaborating.  It may be that all our goals are somewhat different, but it's worth a discussion (IMO).

Regarding your "-Y" option, and querying before hardlinking, I had a similar thought a while back, because I generally found myself wanting to use dry-run for nearly all the runs (and looking at the reported statistics before making a final decision on whether to hardlink), so I made my variant essentially default to "dry-run" mode, and require an option to actually perform linking.  In general, I tend to prefer the "safe" default for these things.  My ambitious idea, possibly overkill, is to simultaneously gather and compute accurate statistics for all option variants simultaneously (ie. still with a single tree walk), so that it's possible to choose which kind of hardlinking is desired (ie. "timestamp" or "notimestamp", "ignoring permissions" or not, etc) after just one run rather than repeated runs.  Still very much in the planning stages, however.

Regarding zero-size files, they crop up surprisingly often in my experience, and I think the default of *not* linking them is the right approach, since it is probably nearly always undesired (and unintended) behavior.  The space saving is probably non-existent, since AFAIK most filesystems do not reclaim unused inode space anyway (ie. the data structure is still using disk space, even when unused); perhaps some modern filesystems are different, since we live in an age of Terabyte and even Petabyte filesystems, where reclaiming inode space could possibly matter.  But from a semantic standpoint, linking unrelated zero-length files usually seems like a mistake.

wolfospealain commented:
I don't think my code had become that separate and is still very much the same program but the restructuring as OO has meant a significant single change that cannot be broken into distinct minor changes. The functionality is the same and the basic algorithm is the same but for the introduction of a file object to store inode attributes and related filenames - allowing for backtracking when an inode with more links is discovered. I wouldn't like to imply that mine is a separate program. The code did need to become properly object oriented if it was to continue to develop. I dislike Python 2 only because of its weird print statement - second only to the ugliness of elif. I use the program myself and just primarily I needed to solve the memory problem of large amounts of files and the annoying clustering of hardlinks. It will never be fully resolved as rsync will still introduce into the backup new files which are identical files with just different attributes. So back to running hardlink again.

akaihola commented on hardlink.py:
```diff
@@ -58,26 +58,29 @@ def hardlink(self, other, dry_run=False, verbose=0):
                     # hardlink succeeded, update links
-                    source.files.update({link: (destination.files[link][0], 1)})
                     source.files.update({source.path: (source.files[source.path][0], source.files[source.path][1] + 1)})
+                    logging.debug("SOURCE " + source.path + " " +str(source.files[source.path][0]) + ", " +str(source.files[source.path][1]))
```
For clarity, I would make use of the logging framework's string formatting support and rewrite these as:

```python
    logging.debug('SOURCE %s %s, %s',
                  source.path,
                  source.files[source.path][0],
                  source.files[source.path][1])
    logging.debug('DESTINATION %s %d, 1',
                  link,
                  destination.files[link][0] - 1)
```

akaihola commented on hardlink.py:
```diff
@@ -58,26 +58,29 @@ def hardlink(self, other, dry_run=False, verbose=0):
                     source.files.update({source.path: (source.files[source.path][0], source.files[source.path][1] + 1)})
+                    logging.debug("SOURCE " + source.path + " " +str(source.files[source.path][0]) + ", " +str(source.files[source.path][1]))
+                    logging.debug("DESTINATION " + link + " " +str(destination.files[link][0] - 1) + ", 1")
+                    source.files.update({link: (destination.files[link][0] - 1, 1)})
```
It would probably be more idiomatic to just write
```python
source.files[link] = destination.files[link][0] - 1, 1
```

akaihola commented:
@wolfospealain I can see how the OO rewrite is necessary. It's a big structural change, and in order for collaboration to be possible and the history of the code base to be clean, it should in my opinion happen as purely a refactoring which causes no functional changes. Preferably our main fork should be stable with no open pull requests before the OO rewrite or during its review.

chadnetzer commented:
@wolfospealain Re: "I wouldn't like to imply that mine is a separate program", the LICENSE change alone to GPL 3 implies complications with backporting or merging any changes to the @akaihola master, as it'll have to either be dual licensed (ie. the terms would require having both the GPL 2 and 3 license files present in this repo), or @akaihola would have to choose to relicense to GPLv3.  Beyond that to technical issues, there is essentially no possibility of a meaningful "merge" of pieces back to the current code (imo), without simply replacing it all (an issue shared with my current development HEAD).  So there is a strong likelihood of your repo remaining a separate program.  Or at the least, the issue of merging and integrating your changes is decidedly more of a non-technical one.

chadnetzer commented:
@wolfospealain On a technical matter, I was curious if the chown() you do to the owner of the most recently modified inode is necessary; ie. how does it change the behavior of the rsync backups if the chown() didn't happen?  Because, as I see it, when linking two files of different ownership, necessarily one is going to change, and it seems somewhat arbitrary to change ownership to the most recently modified version; though there is some logic to it, it's still somewhat arbitrary and it seems to me that it'll trigger backups to re-backup one of the files anyway.  Since os.chown() is documented as Unix only, and (typically) requires running as root, imo it'd be nice to avoid if not necessary.

chadnetzer commented:
@akaihola Re: "the OO rewrite is necessary... it should in my opinion happen as purely a refactoring which causes no functional changes.", I think that ship may have sailed, practically speaking.  For me at least, I might still be willing to invest the effort, but for example, my PR #33 makes a lot of internal and feature changes that are built on top of the current mostly-non-OO master, before I then refactored into a fully OO codebase; @wolfospealain presumably had a similar issue, where at a certain point it became too difficult to put off the refactoring.  But in both cases, to refactor the current master without those intervening functional changes will imply a *lot* of effort to reimplement them on our part (if indeed either of our code bases are desirable as the future direction to take this repo).

It's just a side effect of the recent and rapid nature of both of our forked repos having undergone changes, after a somewhat long dormant period; there are dozens to hundreds of commits worth of changes to consider, and it just may not be practical to rewrite them (imo).  That said, if you have a particular clear vision for how you want the refactoring to proceed, I'm open to hearing it.  Best to move that discussion to issue #36 however.

wolfospealain committed:
Update README.md

wolfospealain committed:
Update README.md

wolfospealain committed:
Update README.md

wolfospealain committed:
Statistics, confirmation option, zero length files.

Improved statistics and reporting. Add -Y confirmation option. Excludes zero length files.

wolfospealain committed:
Update _tests.py

wolfospealain committed:
Statistics and reporting upgrade.

wolfospealain committed:
Update README.md

wolfospealain committed:
Update README.md

wolfospealain committed:
Persistent database file for testing and incremental passes.

wolfospealain committed:
Update hardlink.py

wolfospealain committed:
Update README.md

wolfospealain committed:
Persistent database option.

Persistent database option for testing, dry-run data collection, and incremental scans.

wolfospealain committed:
Avoid re-comparing known inodes this runtime. Correct dry-run statistics.

wolfospealain committed:
Update README.md

wolfospealain committed:
A hardlink command that works.
